### PR TITLE
Move all regenerate? checking to Regenerator

### LINKED ISSUE #3311: undefined method `regenerate?'
@envygeeks @alfredxing - Thanks for the help and getting incremental build (https://github.com/jekyll/jekyll/pull/3116)  in!

I've been testing out incremental regeneration on latest. Couple of liquid 3.0 fixes, no problems with that. But I did run into an issue when trying to serve it up, any ideas? Here's the stack trace:

```
Configuration file: /home/eric/workspace/website/_config.yml
            Source: src
       Destination: /home/eric/workspace/website/_site
 Incremental build: enabled
      Generating... 
/home/eric/.rbenv/versions/2.0.0-p481/lib/ruby/gems/2.0.0/gems/jekyll-2.5.3/lib/jekyll/site.rb:322:in `block in write': undefined method `regenerate?' for #<Sprockets::StaticAsset:0x007f662353c5d8> (NoMethodError)
	from /home/eric/.rbenv/versions/2.0.0-p481/lib/ruby/gems/2.0.0/gems/jekyll-2.5.3/lib/jekyll/site.rb:481:in `block (2 levels) in each_site_file'
	from /home/eric/.rbenv/versions/2.0.0-p481/lib/ruby/gems/2.0.0/gems/jekyll-2.5.3/lib/jekyll/site.rb:480:in `each'
	from /home/eric/.rbenv/versions/2.0.0-p481/lib/ruby/gems/2.0.0/gems/jekyll-2.5.3/lib/jekyll/site.rb:480:in `block in each_site_file'
	from /home/eric/.rbenv/versions/2.0.0-p481/lib/ruby/gems/2.0.0/gems/jekyll-2.5.3/lib/jekyll/site.rb:479:in `each'
	from /home/eric/.rbenv/versions/2.0.0-p481/lib/ruby/gems/2.0.0/gems/jekyll-2.5.3/lib/jekyll/site.rb:479:in `each_site_file'
	from /home/eric/.rbenv/versions/2.0.0-p481/lib/ruby/gems/2.0.0/gems/jekyll-2.5.3/lib/jekyll/site.rb:321:in `write'
	from /home/eric/.rbenv/versions/2.0.0-p481/lib/ruby/gems/2.0.0/gems/jekyll-assets-0.9.2/lib/jekyll/assets_plugin/patches/site_patch.rb:61:in `__wrap_write'
	from /home/eric/.rbenv/versions/2.0.0-p481/lib/ruby/gems/2.0.0/gems/jekyll-2.5.3/lib/jekyll/site.rb:57:in `process'
	from /home/eric/.rbenv/versions/2.0.0-p481/lib/ruby/gems/2.0.0/gems/jekyll-2.5.3/lib/jekyll/command.rb:28:in `process_site'
	from /home/eric/.rbenv/versions/2.0.0-p481/lib/ruby/gems/2.0.0/gems/jekyll-2.5.3/lib/jekyll/commands/build.rb:58:in `build'
	from /home/eric/.rbenv/versions/2.0.0-p481/lib/ruby/gems/2.0.0/gems/jekyll-2.5.3/lib/jekyll/commands/build.rb:34:in `process'
	from /home/eric/.rbenv/versions/2.0.0-p481/lib/ruby/gems/2.0.0/gems/jekyll-2.5.3/lib/jekyll/commands/serve.rb:26:in `block (2 levels) in init_with_program'
	from /home/eric/.rbenv/versions/2.0.0-p481/lib/ruby/gems/2.0.0/gems/mercenary-0.3.5/lib/mercenary/command.rb:220:in `call'
	from /home/eric/.rbenv/versions/2.0.0-p481/lib/ruby/gems/2.0.0/gems/mercenary-0.3.5/lib/mercenary/command.rb:220:in `block in execute'
	from /home/eric/.rbenv/versions/2.0.0-p481/lib/ruby/gems/2.0.0/gems/mercenary-0.3.5/lib/mercenary/command.rb:220:in `each'
	from /home/eric/.rbenv/versions/2.0.0-p481/lib/ruby/gems/2.0.0/gems/mercenary-0.3.5/lib/mercenary/command.rb:220:in `execute'
	from /home/eric/.rbenv/versions/2.0.0-p481/lib/ruby/gems/2.0.0/gems/mercenary-0.3.5/lib/mercenary/program.rb:42:in `go'
	from /home/eric/.rbenv/versions/2.0.0-p481/lib/ruby/gems/2.0.0/gems/mercenary-0.3.5/lib/mercenary.rb:19:in `program'
	from /home/eric/.rbenv/versions/2.0.0-p481/lib/ruby/gems/2.0.0/gems/jekyll-2.5.3/bin/jekyll:17:in `<top (required)>'
	from /home/eric/.rbenv/versions/2.0.0-p481/bin/jekyll:23:in `load'
	from /home/eric/.rbenv/versions/2.0.0-p481/bin/jekyll:23:in `<main>'
```

### PR DESCRIPTION
**Pull request by alfredxing**
Moves all Post/Page/Document `regenerate?` checking to Regenerator (renamed from Metadata). Also allows non-core objects to be rendered with out an error (fixes #3311).

/cc @parkr @envygeeks @crearc

### CODE REVIEW

**Comment by parkr** [comment]
Cursory glance says :+1: . Will take a closer look tonight. Thank you!!

File: lib/jekyll/site.rb [hunk 1]
```diff
@@ -319,9 +323,9 @@ def cleanup
     # Returns nothing.
     def write
       each_site_file { |item|
-        item.write(dest) if item.regenerate?
+        item.write(dest) if regenerator.regenerate?(item)
       }
-      metadata.write unless full_rebuild?
+      regenerator.write unless full_rebuild?
```
Comment (parkr) [inline lib/jekyll/site.rb]: This method is kind of ambiguous.

File: test/test_regenerator.rb [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+
+      @site = Site.new(Jekyll.configuration({
+        "source" => source_dir,
+        "destination" => dest_dir,
+        "collections" => {
+          "methods" => {
+            "output" => true
+          }
+        }
+      }))
```
Comment (parkr) [inline test/test_regenerator.rb]: I added a `fixture_site` somewhere or have a means of generating this without touching the file system. Somewhere in `test/helper.rb`. Could you please use that?

File: lib/jekyll/site.rb (hunk 1 above)
Comment (alfredxing) [inline lib/jekyll/site.rb]: The naming?

Comment (parkr) [inline lib/jekyll/site.rb]: Yeah!

Comment (alfredxing) [inline lib/jekyll/site.rb]: Alright, fixed. It was `metadata.write` before, which made more sense, but now it's renamed to `regenerator.write_metadata`.

**Comment by parkr** [comment]
LGTM. @envygeeks / @mattr- , wanna take a look?

**Comment by envygeeks** [comment]
:+1:

**Comment by parkr** [comment]
Thanks!!

**Pull request merged**